Linear solver backward

mh-nguyen712 opened the pull request:
Thank you for contributing to DeepInverse!

Write custom backward function for `least_squares` solver:
- Can differentiate w.r.t all input and physics parameters
- Less memory consumption
- Sometimes faster than backprop
- But not support TensorList  

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

jscanvic commented on deepinv/tests/test_optim.py:
```diff
@@ -985,3 +987,106 @@ def A_adjoint(self, y, **kwargs):
     assert rel_error < 0.1
+
+
+from deepinv.optim.utils import least_squares_implicit_backward
```
Move imports to the top

jscanvic commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +775,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+
+        kwargs = extra_kwargs if extra_kwargs is not None else {}
+
+        with torch.no_grad():
```
Is it needeed? I feel like the torch engine should run autograd Function code in no_grad by default

jscanvic commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +775,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+                **kwargs,
+            )
+        # Save tensors only
+        gamma = torch.as_tensor(gamma, dtype=y.dtype, device=y.device)
```
If possible: don't convert `gamma` to a tensor and save it in `ctx.gamma = gamma` instead of in the `ctx.save_for_backward` call

jscanvic commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +775,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+        physics = ctx.physics
+
+        # Solve (A^T A + I/gamma) mv = grad_output
+        with torch.no_grad():
```
Is no_grad needed here as well?

jscanvic commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +775,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+        needs = ctx.needs_input_grad
+        grads = [None] * len(needs)
+
+        # dL/dy = A mv
```
I'm not sure we should need to do much more than running `least_squares` here. It looks like we're computing backwards of related nodes manually but I'm pretty sure we could rely on existing backwards implementations somehow, with a loop or something. Thoughts?

jscanvic reviewed (changes requested):
Thanks Hai for this really nice PR, it looks really promising! Some comments

tachella reviewed (changes requested):
I think this is adding really nice functionalities! I added some suggestions for the docs, I think we can give much more details.

I also realised that we were testing the wrong operators here

https://github.com/deepinv/deepinv/blob/eeae34cc57a29c587dbce54fab9c62717c71e63b/deepinv/tests/test_optim.py#L883C40-L883C50

since most of them have SVD decompositions, and the least squares solver is not called!

mh-nguyen712 commented:
Thanks for your review. I just updated the examples and the test (I changed a bit a the test and lower tolerance). I also fix the first issue in #769.

tachella commented on deepinv/physics/forward.py:
```diff
@@ -355,6 +356,7 @@ class LinearPhysics(Physics):
     :param float tol: If the operator does not have a closed form pseudoinverse, a least squares algorithm
         is used for computing it, and this parameter fixes the relative tolerance of the least squares algorithm.
     :param str solver: least squares solver to use. Choose between `'CG'`, `'lsqr'`, `'BiCGStab'` and `'minres'`. See :func:`deepinv.optim.utils.least_squares` for more details.
+    :param bool implicit_backward_solver: If `True`, uses implicit differentiation for computing gradients through the :meth:`deepinv.physics.LinearPhysics.A_dagger` and :meth:`deepinv.physics.LinearPhysics.prox_l2`, using :func:`deepinv.optim.utils.least_squares_implicit_backward` instead of :func:`deepinv.optim.utils.least_squares`. This can significantly reduce memory consumption, especially w ...
```
```suggestion
    :param bool implicit_backward_solver: If `True`, uses implicit differentiation for computing gradients through the :meth:`deepinv.physics.LinearPhysics.A_dagger` and :meth:`deepinv.physics.LinearPhysics.prox_l2`, using :func:`deepinv.optim.utils.least_squares_implicit_backward` instead of :func:`deepinv.optim.utils.least_squares`. This can significantly reduce memory consumption, especially when using many iterations. If `False`, uses the standard autograd mechanism, which can be memory-intensive. Default is `True`.
```

tachella commented on deepinv/tests/test_optim.py:
```diff
@@ -985,3 +987,116 @@ def A_adjoint(self, y, **kwargs):
+    buffers = copy.deepcopy(dict(physics.named_buffers()))
+    parameters = {k: v for k, v in buffers.items() if k in params}
+    for k, v in parameters.items():
+        if isinstance(v, torch.Tensor) and v.dtype in [
```
maybe add a comment explaining that we do this to only backprop wrt to continuous variables

mh-nguyen712 replied:
done

tachella commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+# ----------------------------------------------------------------------------------------
+#
+
+BASE_DIR = Path(".")
```
these 2 lines are not used I believe

tachella commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+import matplotlib.pyplot as plt
+
+# %%
+# Setup paths for data loading and results.
```
I guess we can remove this subsection if we remove the 2 lines below

tachella commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+# In this example, we use the CBSD500 dataset for training and the Set3C dataset for testing.
+
+img_size = 64 if torch.cuda.is_available() else 32
+n_channels = 3  # 3 for color images, 1 for gray-scale images
```
this is not used

tachella commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+# %%
+# Generate a dataset of low resolution images and load it.
+# ----------------------------------------------------------------------------------------
+# We use the Blur class from the physics module to generate a dataset of blurry images.
```
maybe we can explain that we choose blur with 'valid' padding since it its proximal operator does not have a closed form solution and requires using a least squares solver

tachella commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+.. math::
+
+    \left( \frac{\partial h}{\partial z} \right)^T v               &= \frac{1}{\gamma} M v \\
+    \left( \frac{\partial h}{\partial y} \right)^T v               &= A_\theta M v \\
```
```suggestion
    \left( \frac{\partial h}{\partial y} \right)^{\top} v               &= A_\theta M v \\
```

tachella reviewed (changes requested):
many thanks for all the updates!

I have some additional comments, but I think we should converge soon :)

mh-nguyen712 commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +789,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+    .. tip::
+
+        Training unfolded network with implicit differentiation can reduce memory consumption significantly, especially when using many iterations.
+        TODO: link to example notebook when available.
```
done

mh-nguyen712 commented on examples/unfolded/demo_unfolded_constant_memory.py:
```diff
@@ -0,0 +1,277 @@
+    )
+
+
+# Compare the training loss
```
- Yes the blur curve is just behind the orange one. I adjust the linestyle to see it better
- I increase the number of iteration to see a time difference (~1.1x-1.2x speed up, not much). The most interesting part is 2x-3x memory reduction factor, but I can only trace these info in GPU.

mh-nguyen712 replied:
I just added a note at the beginning on the example and in the function as well

tachella replied:
perfect, thank you!

mh-nguyen712 commented on deepinv/optim/utils.py:
```diff
@@ -769,6 +771,221 @@ def gradient_descent(grad_f, x, step_size=1.0, max_iter=1e2, tol=1e-5):
+
+        Training unfolded network with implicit differentiation can reduce memory consumption significantly, especially when using many iterations.
+        TODO: link to example notebook when available.
+
```
Done
